firedrake/assemble.py

```diff
@@ -222,32 +225,42 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
+        # bottom and interior horizontal facets will use the cell to dofs map
+        # coming from the base mesh as a starting point for the actual dynamic map
+        # computation.
+        def integral_domains(integral_type):
+            if integral_type == "cell":
+                cell_domains.append(op2.ALL)
+            elif integral_type == "exterior_facet":
+                exterior_facet_domains.append(op2.ALL)
+            elif integral_type == "interior_facet":
+                interior_facet_domains.append(op2.ALL)
+            elif integral_type == "exterior_facet_bottom":
+                cell_domains.append(op2.ON_BOTTOM)
+            elif integral_type == "exterior_facet_top":
+                cell_domains.append(op2.ON_TOP)
+            elif integral_type == "exterior_facet_vert":
+                exterior_facet_domains.append(op2.ALL)
+            elif integral_type == "interior_facet_horiz":
+                cell_domains.append(op2.ON_INTERIOR_FACETS)
+            elif integral_type == "interior_facet_vert":
+                interior_facet_domains.append(op2.ALL)
+            else:
+                raise ValueError('Unknown integral type "%s"' % integral_type)
+
         if tensor is None:
-            # For horizontal facets of extrded meshes, the corresponding domain
-            # in the base mesh is the cell domain. Hence all the maps used for top
-            # bottom and interior horizontal facets will use the cell to dofs map
-            # coming from the base mesh as a starting point for the actual dynamic map
-            # computation.
-            for integral in integrals:
-                integral_type = integral.integral_type()
-                if integral_type == "cell":
-                    cell_domains.append(op2.ALL)
-                elif integral_type == "exterior_facet":
-                    exterior_facet_domains.append(op2.ALL)
-                elif integral_type == "interior_facet":
-                    interior_facet_domains.append(op2.ALL)
-                elif integral_type == "exterior_facet_bottom":
-                    cell_domains.append(op2.ON_BOTTOM)
-                elif integral_type == "exterior_facet_top":
-                    cell_domains.append(op2.ON_TOP)
-                elif integral_type == "exterior_facet_vert":
-                    exterior_facet_domains.append(op2.ALL)
-                elif integral_type == "interior_facet_horiz":
-                    cell_domains.append(op2.ON_INTERIOR_FACETS)
-                elif integral_type == "interior_facet_vert":
-                    interior_facet_domains.append(op2.ALL)
-                else:
-                    raise ValueError('Unknown integral type "%s"' % integral_type)
+            # SLATE expressions do not contain integrals. Therefore, we loop over the kernels
+            # in order to determine integral domain information. Otherwise, we loop over the
+            # integrals of the ufl form.
+            if is_slate_expr:
```

Comment:
So I'm not quite sure this is a good solution. The reason I added this is because SLATE objects do not have an `integral` method. All the kernels do have an `integral_type` associated with it though, so I loop over kernels. This approach works for nearly all cases of `form` objects as well. However, if a kernel evaluates to zero, tsfc doesn't return a kernel (I believe this is related to the `NoopError` in tsfc?). Is there a better solution or does TSFC need to be modified?